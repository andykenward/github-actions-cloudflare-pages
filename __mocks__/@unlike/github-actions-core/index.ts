import * as core from '@unlike/github-actions-core'
import {vi} from 'vitest'

export const getInput = vi.fn<
  Parameters<typeof core.getInput>,
  ReturnType<typeof core.getInput>
>((name: string, options) => core.getInput(name, options))

export const setOutput = vi.fn()

export const error = vi.fn()

export const summary = core.summary
summary.addTable = vi.fn().mockReturnValue(summary)
summary.addHeading = vi.fn().mockReturnValue(summary)
summary.addBreak = vi.fn().mockReturnValue(summary)
summary.write = vi.fn()
