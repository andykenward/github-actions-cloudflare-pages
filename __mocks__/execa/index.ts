import {vi} from 'vitest'

export const $ = vi.fn()
