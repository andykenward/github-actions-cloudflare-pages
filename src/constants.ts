/**
 * GitHub Action Input
 */

/** GitHub Action Input key - Cloudflare Account Id */
export const ACTION_INPUT_ACCOUNT_ID = 'account id'
/** GitHub Action Input key - Cloudflare Pages Project Name */
export const ACTION_INPUT_PROJECT_NAME = 'project name'
/** GitHub Action Input key - Cloudflare API token */
export const ACTION_INPUT_API_TOKEN = 'api token'
/** GitHub Action Input key = Directory of static files to upload */
export const ACTION_INPUT_DIRECTORY = 'directory'
export const ACTION_INPUT_GITHUB_TOKEN = 'github token'
export const ACTION_INPUT_GITHUB_ENVIRONMENT = 'github environment'
/**
 * Environment variables
 */

/** Environment variables key for Cloudflare wrangler */
export const CLOUDFLARE_API_TOKEN = 'CLOUDFLARE_API_TOKEN'
/** Environment variables key for Cloudflare wrangler */
export const CLOUDFLARE_ACCOUNT_ID = 'CLOUDFLARE_ACCOUNT_ID'
