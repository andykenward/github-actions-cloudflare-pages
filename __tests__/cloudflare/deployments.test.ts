import {setOutput, summary} from '@unlike/github-actions-core'
import * as execa from 'execa'
import {afterEach, beforeEach, describe, expect, test, vi} from 'vitest'

import type {PagesDeployment} from '@/src/cloudflare/types.js'
import type {MockApi} from '@/tests/helpers/index.js'
import RESPONSE_NOT_FOUND_DEPLOYMENTS from '@/responses/api.cloudflare.com/pages/deployments/deployments-not-found.response.json'
import RESPONSE_DEPLOYMENTS from '@/responses/api.cloudflare.com/pages/deployments/deployments.response.json'
import {
  createDeployment,
  getDeploymentAlias
} from '@/src/cloudflare/deployments.js'
import {CLOUDFLARE_ACCOUNT_ID, CLOUDFLARE_API_TOKEN} from '@/src/constants.js'
import {
  MOCK_API_PATH_DEPLOYMENTS,
  setMockApi,
  stubRequiredInputEnv
} from '@/tests/helpers/index.js'

import PACKAGE_JSON from '../../package.json'

vi.mock('execa')
vi.mock('@unlike/github-actions-core')
describe('deployments', () => {
  describe('getDeploymentAlias', () => {
    test('returns deployment url when aliases is null', () => {
      const alias = getDeploymentAlias({
        aliases: null,
        url: 'https://helloworld.pages.dev'
      } as PagesDeployment)

      expect(alias).toStrictEqual('https://helloworld.pages.dev')
    })

    test('returns deployment url when aliases is empty array', () => {
      const alias = getDeploymentAlias({
        aliases: [],
        url: 'https://helloworld.pages.dev'
      } as unknown as PagesDeployment)

      expect(alias).toStrictEqual('https://helloworld.pages.dev')
    })

    test('returns deployment first alias when aliases is not empty array', () => {
      const alias = getDeploymentAlias({
        aliases: ['https://helloworld-alias.pages.dev'],
        url: 'https://helloworld.pages.dev'
      } as PagesDeployment)

      expect(alias).toStrictEqual('https://helloworld-alias.pages.dev')
    })
  })

  describe('createDeployment', () => {
    describe('api calls', () => {
      let mockApi: MockApi
      beforeEach(() => {
        mockApi = setMockApi()
        // Set required inputs
        stubRequiredInputEnv()
      })
      afterEach(async () => {
        mockApi.mockAgent.assertNoPendingInterceptors()
        await mockApi.mockAgent.close()
      })

      test('handles thrown error from wrangler deploy', async () => {
        expect.assertions(9)
        // Mock $ execa rejects
        vi.mocked(execa.$).mockRejectedValue({stderr: 'Oh no!'})

        // Expect Cloudflare Api Token and Account Id to be undefined.
        expect(process.env[CLOUDFLARE_API_TOKEN]).toBeUndefined()
        expect(process.env[CLOUDFLARE_ACCOUNT_ID]).toBeUndefined()

        await expect(
          createDeployment()
        ).rejects.toThrowErrorMatchingInlineSnapshot('"Oh no!"')

        expect(execa.$).toHaveBeenCalledWith(
          [
            `npx wrangler@${PACKAGE_JSON.dependencies.wrangler} pages deploy `,
            ' --project-name=',
            ' --branch=',
            ' --commit-dirty=true --commit-hash=',
            ''
          ],
          'mock-directory',
          'mock-project-name',
          'mock-github-head-ref',
          'mock-github-sha'
        )
        expect(execa.$).toHaveBeenCalledTimes(1)
        expect(process.env[CLOUDFLARE_API_TOKEN]).toStrictEqual(
          'mock-api-token'
        )
        expect(process.env[CLOUDFLARE_ACCOUNT_ID]).toStrictEqual(
          'mock-account-id'
        )
        expect(setOutput).not.toHaveBeenCalled()
        expect(summary.addTable).not.toHaveBeenCalled()
      })

      test('handles thrown error from getDeployments', async () => {
        expect.assertions(4)
        vi.mocked(execa.$).mockResolvedValue({
          isCanceled: false,
          command: '',
          escapedCommand: '',
          exitCode: 0,
          stdout: 'success',
          stderr: '',
          failed: false,
          timedOut: false,
          killed: false
        })

        mockApi.interceptCloudflare(
          MOCK_API_PATH_DEPLOYMENTS,
          RESPONSE_NOT_FOUND_DEPLOYMENTS,
          404
        )

        await expect(
          createDeployment()
        ).rejects.toThrowErrorMatchingInlineSnapshot(
          '"A request to the Cloudflare API (https://api.cloudflare.com/client/v4/accounts/mock-account-id/pages/projects/mock-project-name/deployments) failed."'
        )
        expect(execa.$).toHaveBeenCalledTimes(1)
        expect(setOutput).not.toHaveBeenCalled()
        expect(summary.addTable).not.toHaveBeenCalled()
      })

      test('handles success', async () => {
        expect.assertions(11)
        vi.mocked(execa.$).mockResolvedValue({
          isCanceled: false,
          command: '',
          escapedCommand: '',
          exitCode: 0,
          stdout: 'success',
          stderr: '',
          failed: false,
          timedOut: false,
          killed: false
        })

        mockApi.interceptCloudflare(
          MOCK_API_PATH_DEPLOYMENTS,
          RESPONSE_DEPLOYMENTS,
          200
        )

        const deployment = await createDeployment()

        expect(deployment).toMatchSnapshot()
        expect(deployment.id).toMatchInlineSnapshot(
          '"206e215c-33b3-4ce4-adf4-7fc6c9b65483"'
        )

        expect(setOutput).toHaveBeenCalledTimes(4)
        expect(setOutput).toHaveBeenNthCalledWith(
          1,
          'id',
          '206e215c-33b3-4ce4-adf4-7fc6c9b65483'
        )
        expect(setOutput).toHaveBeenNthCalledWith(
          2,
          'url',
          'https://206e215c.cloudflare-pages-action-a5z.pages.dev'
        )
        expect(setOutput).toHaveBeenNthCalledWith(
          3,
          'environment',
          'production'
        )
        expect(setOutput).toHaveBeenNthCalledWith(
          4,
          'alias',
          'https://unknown-branch.cloudflare-pages-action-a5z.pages.dev'
        )

        expect(summary.addHeading).toHaveBeenCalledWith(
          `Cloudflare Pages Deployment`
        )
        expect(summary.addBreak).toHaveBeenCalledTimes(1)

        expect(summary.addTable).toHaveBeenCalledTimes(1)
        expect(summary.addTable).toHaveBeenCalledWith([
          [
            {
              data: 'Name',
              header: true
            },
            {
              data: 'Result',
              header: true
            }
          ],
          ['Environment:', `production`],
          [
            'Branch:',
            `<a href='https://github.com/unlike-ltd/github-actions-cloudflare-pages/tree/main'><code>main</code></a>`
          ],
          [
            'Commit Hash:',
            `<a href='https://github.com/unlike-ltd/github-actions-cloudflare-pages/commit/mock-github-sha'><code>mock-github-sha</code></a>`
          ],
          ['Commit Message:', `chore(deps-dev): update eslint packages`],
          ['Status:', `<strong>SUCCESS</strong>`],
          [
            'Preview URL:',
            `<a href='https://206e215c.cloudflare-pages-action-a5z.pages.dev'>https://206e215c.cloudflare-pages-action-a5z.pages.dev</a>`
          ],
          [
            'Branch Preview URL:',
            `<a href='https://unknown-branch.cloudflare-pages-action-a5z.pages.dev'>https://unknown-branch.cloudflare-pages-action-a5z.pages.dev</a>`
          ]
        ])
      })
    })
  })
})
