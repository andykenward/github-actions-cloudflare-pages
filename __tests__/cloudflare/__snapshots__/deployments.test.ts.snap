// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deployments > createDeployment > api calls > handles success 1`] = `
{
  "aliases": [
    "https://unknown-branch.cloudflare-pages-action-a5z.pages.dev",
  ],
  "build_config": {
    "build_command": null,
    "destination_dir": null,
    "root_dir": null,
    "web_analytics_tag": null,
    "web_analytics_token": null,
  },
  "build_image_major_version": 1,
  "compatibility_date": "2023-06-02",
  "compatibility_flags": [],
  "created_on": "2023-06-16T18:41:19.07168Z",
  "deployment_trigger": {
    "metadata": {
      "branch": "main",
      "commit_dirty": true,
      "commit_hash": "mock-github-sha",
      "commit_message": "chore(deps-dev): update eslint packages",
    },
    "type": "ad_hoc",
  },
  "env_vars": {},
  "environment": "production",
  "id": "206e215c-33b3-4ce4-adf4-7fc6c9b65483",
  "is_skipped": false,
  "latest_stage": {
    "ended_on": "2023-06-16T18:41:21.226094Z",
    "name": "deploy",
    "started_on": null,
    "status": "success",
  },
  "modified_on": "2023-06-16T18:41:21.226094Z",
  "production_branch": "main",
  "project_id": "48ddb51b-4568-4f38-94de-10614c2fa270",
  "project_name": "cloudflare-pages-action",
  "short_id": "206e215c",
  "stages": [
    {
      "ended_on": null,
      "name": "queued",
      "started_on": "2023-06-16T18:41:19.07168Z",
      "status": "active",
    },
    {
      "ended_on": null,
      "name": "initialize",
      "started_on": null,
      "status": "idle",
    },
    {
      "ended_on": null,
      "name": "clone_repo",
      "started_on": null,
      "status": "idle",
    },
    {
      "ended_on": null,
      "name": "build",
      "started_on": null,
      "status": "idle",
    },
    {
      "ended_on": "2023-06-16T18:41:21.226094Z",
      "name": "deploy",
      "started_on": null,
      "status": "success",
    },
  ],
  "url": "https://206e215c.cloudflare-pages-action-a5z.pages.dev",
  "usage_model": null,
}
`;
